lib/cache/inventory/inventory_cache.go

```diff
@@ -600,34 +752,324 @@ func (ic *InventoryCache) ListUnifiedInstances(ctx context.Context, req *invento
+type unifiedFilterEnvironment struct {
+	ui *inventoryInstance
+}
+
+func (e *unifiedFilterEnvironment) GetVersion() string {
+	if e == nil || e.ui == nil {
+		return ""
+	}
+	if e.ui.isInstance() {
+		return e.ui.instance.Spec.Version
+	}
+	// For bot instances, get version from latest heartbeat
+	if e.ui.bot.Status != nil && len(e.ui.bot.Status.LatestHeartbeats) > 0 {
+		return e.ui.bot.Status.LatestHeartbeats[0].Version
+	}
+	return ""
+}
+
+func (e *unifiedFilterEnvironment) GetHostname() string {
+	if e == nil || e.ui == nil {
+		return ""
+	}
+	if e.ui.isInstance() {
+		return e.ui.instance.Spec.Hostname
+	}
+	// For bot instances, get hostname from latest heartbeat
+	if e.ui.bot.Status != nil && len(e.ui.bot.Status.LatestHeartbeats) > 0 {
+		return e.ui.bot.Status.LatestHeartbeats[0].Hostname
+	}
+	return ""
+}
+
+func (e *unifiedFilterEnvironment) GetName() string {
+	if e == nil || e.ui == nil {
+		return ""
+	}
+	if e.ui.isInstance() {
+		return e.ui.instance.GetMetadata().Name
+	}
+	return e.ui.bot.GetMetadata().GetName()
+}
+
+func (e *unifiedFilterEnvironment) GetBotName() string {
+	if e == nil || e.ui == nil || e.ui.isInstance() {
+		return ""
+	}
+	return e.ui.bot.Spec.BotName
+}
+
+func (e *unifiedFilterEnvironment) GetInstanceID() string {
+	if e == nil || e.ui == nil {
+		return ""
+	}
+	if e.ui.isInstance() {
+		return e.ui.instance.GetName()
+	}
+	return e.ui.bot.Spec.InstanceId
+}
+
+func (e *unifiedFilterEnvironment) GetServices() []string {
+	if e == nil || e.ui == nil || !e.ui.isInstance() {
+		return nil
+	}
+	services := e.ui.instance.Spec.Services
+	result := make([]string, len(services))
+	for i, svc := range services {
+		result[i] = string(svc)
+	}
+	return result
+}
+
+func (e *unifiedFilterEnvironment) GetUpdaterGroup() string {
+	if e == nil || e.ui == nil {
+		return ""
+	}
+	if e.ui.isInstance() {
+		if e.ui.instance.Spec.UpdaterInfo != nil {
+			return e.ui.instance.Spec.UpdaterInfo.UpdateGroup
+		}
+		return ""
+	}
+	// For bot instances, get from latest heartbeat
+	if e.ui.bot.Status != nil && len(e.ui.bot.Status.LatestHeartbeats) > 0 && e.ui.bot.Status.LatestHeartbeats[0].UpdaterInfo != nil {
+		return e.ui.bot.Status.LatestHeartbeats[0].UpdaterInfo.UpdateGroup
+	}
+	return ""
+}
+
+func (e *unifiedFilterEnvironment) GetExternalUpgrader() string {
+	if e == nil || e.ui == nil {
+		return ""
+	}
+	if e.ui.isInstance() {
+		return e.ui.instance.Spec.ExternalUpgrader
+	}
+	// For bot instances, get from latest heartbeat
+	if e.ui.bot.Status != nil && len(e.ui.bot.Status.LatestHeartbeats) > 0 {
+		return e.ui.bot.Status.LatestHeartbeats[0].ExternalUpdater
+	}
+	return ""
+}
+
+// getUnifiedExpressionParser returns a cached parser instance, initializing it once.
+func getUnifiedExpressionParser() (*typical.Parser[*unifiedFilterEnvironment, bool], error) {
+	unifiedExpressionParserOnce.Do(func() {
```

Comment:
I think we'd want to panic on an error (and exercise it in a test so we know that the thing won't actually bomb in the released binary) rather than return an error here - we're fine with `regexp.MustCompile` and that has the same sort of logic.